assets/model_monitoring/components/src/generation_safety_quality/input_schema_adaptor/run.py

```diff
@@ -5,10 +5,55 @@
 
 import argparse
 
+from pyspark.sql import DataFrame
+from pyspark.sql.functions import from_json
 from shared_utilities.io_utils import (
+    init_spark,
     save_spark_df_as_mltable,
     try_read_mltable_in_spark_with_error,
 )
+from shared_utilities.momo_exceptions import InvalidInputError
+from model_data_collector_preprocessor.spark_run import _convert_complex_columns_to_json_string
+from shared_utilities.constants import GENAI_ROOT_SPAN_SCHEMA_COLUMN, GENAI_TRACE_ID_SCHEMA_COLUMN
+
+
+def _adapt_input_data_schema(df: DataFrame) -> DataFrame:
+    """Adapt the input dataframe schema to fit GSQ input schema."""
+    df_field_names = df.schema.fieldNames()
+
+    # check if we need to adapt the schema
+    if GENAI_ROOT_SPAN_SCHEMA_COLUMN not in df_field_names and GENAI_TRACE_ID_SCHEMA_COLUMN not in df_field_names:
+        return df
+
+    spark = init_spark()
+    try:
+        input_schema = spark.read.json(df.rdd.map(lambda row: row.input), mode="FAILFAST").schema
```

Comment:
the input is the trace log of our GenAI preprocessor output, right? I'm thinking may be we can only retrieve the input/output schema from the first, say, 100 rows instead of the whole df to save resources, any concern?

Reply:
I suppose we can take a sampling of the data, it can take a while to read the json. Should we then catch errors thrown later related to the schema when we try unpacking the string with from_json()?

Reply:
Do you think we should use a fraction of rows as opposed to a set number? I was thinking sample 20%